core/services/offchainreporting/contract_tracker.go

```diff
@@ -0,0 +1,329 @@
+
+// OnDisconnect complies with LogListener interface
+func (t *OCRContractTracker) OnDisconnect() {}
+
+// HandleLog complies with LogListener interface
+func (t *OCRContractTracker) HandleLog(lb log.Broadcast, err error) {
+	if err != nil {
+		t.logger.Errorw("OCRContract: error in previous LogListener", "err", err)
+		return
+	}
+
+	was, err := lb.WasAlreadyConsumed()
+	if err != nil {
+		t.logger.Errorw("OCRContract: could not determine if log was already consumed", "error", err)
+		return
+	} else if was {
+		return
+	}
+
+	defer logger.ErrorIfCalling(lb.MarkConsumed)
+
+	topics := lb.RawLog().Topics
+	if len(topics) == 0 {
+		return
+	}
+	raw := lb.RawLog()
+	if raw.Address != t.contractAddress {
+		t.logger.Errorf("log address of 0x%x does not match configured contract address of 0x%x", raw.Address, t.contractAddress)
+		return
+	}
+	switch topics[0] {
+	case OCRContractConfigSet:
+		var configSet *offchainaggregator.OffchainAggregatorConfigSet
+		configSet, err = t.contractFilterer.ParseConfigSet(raw)
+		if err != nil {
+			t.logger.Errorw("could not parse config set", "err", err)
+			return
+		}
+		configSet.Raw = lb.RawLog()
+		cc := confighelper.ContractConfigFromConfigSetEvent(*configSet)
+
+		t.configsMB.Deliver(cc)
+	case OCRContractLatestRoundRequested:
+		var rr *offchainaggregator.OffchainAggregatorRoundRequested
+		rr, err = t.contractFilterer.ParseRoundRequested(raw)
+		if err != nil {
+			t.logger.Errorw("could not parse round requested", "err", err)
+			return
+		}
+		t.lrrMu.Lock()
+		if rr.Round >= t.latestRoundRequested.Round && rr.Epoch >= t.latestRoundRequested.Epoch {
+			t.logger.Infow("OCRContractTracker: received new latest RoundRequested event", "latestRoundRequested", *rr)
+			t.latestRoundRequested = *rr
+		} else {
+			t.logger.Warnw("OCRContractTracker: ignoring out of date RoundRequested event", "latestRoundRequested", t.latestRoundRequested, "roundRequested", rr)
+		}
+		t.lrrMu.Unlock()
+	default:
+		logger.Debugw("OCRContractTracker: got unrecognised log topic", "topic", topics[0])
+	}
+
+}
+
+// IsV2Job complies with LogListener interface
+func (t *OCRContractTracker) IsV2Job() bool {
+	return true
+}
+
+// JobIDV2 complies with LogListener interface
+func (t *OCRContractTracker) JobIDV2() int32 {
+	return t.jobID
+}
+
+// JobID complies with LogListener interface
+func (t *OCRContractTracker) JobID() models.JobID {
+	return models.NilJobID
+}
+
+// SubscribeToNewConfigs returns the tracker aliased as a ContractConfigSubscription
+func (t *OCRContractTracker) SubscribeToNewConfigs(context.Context) (ocrtypes.ContractConfigSubscription, error) {
+	return (*OCRContractConfigSubscription)(t), nil
+}
+
+// LatestConfigDetails queries the eth node
+func (t *OCRContractTracker) LatestConfigDetails(ctx context.Context) (changedInBlock uint64, configDigest ocrtypes.ConfigDigest, err error) {
+	opts := bind.CallOpts{Context: ctx, Pending: false}
+	result, err := t.contractCaller.LatestConfigDetails(&opts)
```

Comment:
Queries the Eth node - is this blocking? Do we need to pass a combined context to the call so it can be cancelled?

Reply:
Yes it queries the eth node and is blocking.

This code is unchanged, it simply moved.

Reply:
Can you elaborate more on what you mean by combined context?

Reply:
We have a pattern in the code, which uses `utils.CombinedContext` to add a "shutdown" channel to a service. This way if you call `Shutdown` any blocking work also stops and it can shut down in a timely manner.

Reply:
Excellent spot, I have added it everywhere in this file.